### DISCUSSION TITLE
React Native Social Login implementation

### DISCUSSION THREAD

**Discussion by Svarto** [nhost/nhost#550] [Tips and Tricks]
This is a short showcase how social login can be implemented in React Native (RN). RN is a bit different to the web, as there are two ways to login through oAuth / Social login, one is the native way and another is using a webView. This implementation uses webView as it is what works on nhost (and it means that the same oAuth implementation would work on both web and mobile).

**Prerequisites:**
- An nhost project with login providers already setup (i.e. Nhost project > Users > Login settings > Social Login Providers
- A React Native mobile app is a working nhost login mechanism (e.g. using onAuthStateChanged to route users into "logged in" area
- I will assume you have a social login button that a user can press, when pressed navigate them to a new screen passing in the "provider" name in the route params (e.g. with react-navigation, `navigation.navigate("SocialLogin", { provider: "google" }).
- This guide uses the package `expo-web-browser` to show a secure login webView on mobile across both Android and iOS. You should have this installed.

**Guide**
Note that I decided to test this against the cloud "prod" environment, as for Google you have to add secure URLs which are allowed to login and they have to have https. It is also possible to funnel your local environment through https://ngrok.com/ if you prefer, and add the funneled endpoint to allowed URLs in your Google oAuth settings.

On the social sign in screen, you will first need to call the `handleSocialLoginAsync` function and when it completes, manage the login status on this page (or any errors):
```
export default function SocialSignInScreen({ navigation, route }) {

    // Deconstruct the social login provider
  const provider = route.params.provider;
  const [result, setResult] = React.useState(null);

  // This manages the result from the social login action (i.e. success, error etc.)
  useEffect(async () => {
    if (result?.type == "cancel") {
      navigation?.goBack();
      toast.show("Login cancelled, please try again", {
        type: "danger",
      });
    }
    if (result?.type == "success") {
      const search = "refreshToken=";
      //Login with
      const session = await nhost.auth.refreshSession(
        result.url.substring(result.url.lastIndexOf(search) + search.length)
      );
      if (session.error) {
        navigation?.goBack();
        toast.show("Error with login, try again or contact support", {
          type: "danger",
        });
      }
    }
  }, [result]);

  // I call this function onLayout, to make sure the graphics are rendered before the
  // oAuth popup shows up for the user
  const handleLogin = async () => {
    const result = await handleSocialLoginAsync(provider);
    setResult(result);
  };
  return (<your social login componen/>)}
```

The handleSocialLoginAsync function, make sure you put in your own nhost url and your app scheme for the redirectURL (mine is yumee://) in the openAuthSessionAsync:

```
import * as React from "react";
import * as WebBrowser from "expo-web-browser";

WebBrowser.maybeCompleteAuthSession();

export const handleSocialLoginAsync = async provider => {
  // Logging into cloud nhost provider
  let result = await WebBrowser.openAuthSessionAsync(
    `https://xxx.nhost.run/v1/auth/signin/provider/${provider}`,
yumee://,
{ preferEphemeralSession: true}
  );
  return result;
};
```

On a successful login in the social login screen will fire the `refreshSession` nhost auth function and as the user is not yet logged in, it will log them in and your normal auth flow will flip to the user as logged in and automatically reroute the user to their logged in area. See above social login screen for where the magic happens.

**Comment by pkreipke** [comment]
@Svarto I'm curious what you used for your NHost based RN template - is there a v2 example/template or guide somewhere that I could follow? The NHost RN example seems to be v1 and 2 yrs old.

**Reply by Svarto** [reply]
Sure! I'll post a new topic on it in tips and tricks. 

Basically, I am using the package `react-navigation` for navigation, and then I have two areas - logged in or not logged in in my routes. Then of course, logged in area is shown if the user is logged in.

The authentication check is executed in a useEffect hook in the topmost app.js file, through the `nhost-js` SDK subscriptions functions (`nhost.auth.isAuthenticated` or `nhost.auth.onAuthStateChange`). These two functions sets a local state variable `loggedIn` to True or False, that in turn switches the user between the mentioned two route categories.

**Comment by jladuval** [comment]
Adding my example here in case it helps people. Just a slight variation on @Svarto but should work to just copy paste direclty
```
import React, { useEffect } from 'react'
import { View } from 'react-native'
import { useNavigation } from '@react-navigation/native'
import * as Linking from 'expo-linking'
import * as WebBrowser from 'expo-web-browser'

import { SUBDOMAIN, nhost } from '../../graphql/nhost' // this is the only line you should need to change

WebBrowser.maybeCompleteAuthSession()

const SocialAuthScreen = ({ route: { params: { provider }}}: { route: { params: { provider: string } } }) =>  {
  const navigation = useNavigation()
  
  useEffect(() => {
    const eff = async () => {
      const result = await WebBrowser.openAuthSessionAsync(
        `https://${SUBDOMAIN}.nhost.run/v1/auth/signin/provider/${provider}`, Linking.createURL('')
      )
      if (result?.type == "cancel" || result?.type === 'dismiss') {
        navigation.goBack()
      }
      if (result?.type == "success") {
        const search = "refreshToken=";
        const session = await nhost.auth.refreshSession(
          result.url.substring(result.url.lastIndexOf(search) + search.length)
        )
        if (session.error) {
          navigation.goBack()
        }
      }
    }
    
    eff()
  }, [])

  return (<View style={{ display: 'flex', flex: 1}}></View>)
}
  
export { SocialAuthScreen }
```

**Reply by jladuval** [reply]
I also added these in my nhost authentication setup (change briefly to the name of your app for deep linking)
![image](https://user-images.githubusercontent.com/1935359/179272694-b0d71688-1bfc-4799-93f7-c9518201a5fe.png)
Follow the instructions here to set up deep linking https://reactnavigation.org/docs/deep-linking/

**Reply by Svarto** [reply]
@jladuval thanks for the nudge, I updated the code slightly in the original post as I made some edits to account for the changes in the latest version of `expo-web-browser`. Specifically added in the redirectURL and also prefer ephemeral session so that on iOS the user does not get a system dialog popup warning that "nhost.run" wants to log them in

**Comment by Danny-A** [comment]
First of all, great work on the initiative!
For me, I am using the Google provider, I had to strip off the last character of the token. The last character is a '#', not sure if this is common from Google.

**Reply by Svarto** [reply]
@Danny-A yes, for Facebook is is _# for some reason. I implemented a stripping function to make sure that I only use the refreshToken.

This started recently, so might change back, so write the stripping function in a robust manner :)

**Comment by ejkkan** [comment]
@Danny-A Would you maybe mind sharing your implementation here as well if it differs from the WebView examples?

Thanks a bunch for this discussion, everyone! It should be highlighted somewhere to spread the knowledge!

**Comment by seyaobey-dev** [comment]
Hi, thank you for putting up this information. I have followed the instructions (created a facebook app and enabled facebook login, also enabled facebook authentication in nhost console). I have followed above steps but I'm still getting the error:
```
URL Blocked: This redirect failed because the redirect URI is not whitelisted...etc
```
I'm not sure what is the redirect url which is blocked? Is it the redirect url found in the `redirect` filed in `nhost dashboard, sign-in methods facebook`? or is it my `app.json scheme: my_app://`. 
Thanks for your assistance!

**Comment by ejkkan** [comment]
@seyaobey-dev 
This setup works for me.
Credentials:
<img width="583" alt="credentials setup" src="https://user-images.githubusercontent.com/32518962/225326616-ca28158d-1891-4a1b-9862-d2ce25b93e03.png">

Nhost google auth config:
<img width="1027" alt="nhostauth" src="https://user-images.githubusercontent.com/32518962/225326677-b73f65f8-bb21-4e7d-9374-c18f83825b92.png">

**Comment by cocodrino** [comment]
Hey guys, sorry for the dumb question. When I run Expo Go, I have the option to open it in my emulator. Should deep linking work in that emulator? I’ve followed the steps—defined the scheme in app.json, installed expo-linking, but when I log this

```tsx
const redirectUrl = Linking.createURL("/");  
console.log("redirectUrl", redirectUrl);  
```

`redirectUrl exp://192.168.0.107:8081/--/`

Is that correct? Also, should I use ngrok or something similar to test it properly?

thanks!
